[TySan] Add initial documentation for Type Sanitizer (#123595)
llvm/llvm-project: clang/docs/TypeSanitizer.rst

================
TypeSanitizer
================

.. contents::
   :local:

Introduction
============

TypeSanitizer is a detector for strict type aliasing violations. It consists of a compiler
instrumentation module and a run-time library. The tool detects violations such as the use 

> of an illegally cast pointer, or misuse of a union.

Reviewer: I think this should say something about accessing memory with a different type than the dynamic type of the object or something like that?

The violations TypeSanitizer catches may cause the compiler to emit incorrect code.

> Typical slowdown introduced by TypeSanitizer is about **4x** [[CHECK THIS]]. Typical memory overhead introduced by TypeSanitizer is about **9x**. 

Reviewer: Might be good to just say that TypeSanitizer is still experimental and currently can have a very large runtime, memory and code size overhead :)

Reviewer 2: It also seems to have a sizable compile-time overhead.

for the use/testing of TypeSanitizer:

.. code-block:: console

   $ cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_RUNTIMES="compiler-rt" <path to source>/llvm

Usage
=====

Compile and link your program with ``-fsanitize=type`` flag.  The
TypeSanitizer run-time library should be linked to the final executable, so
make sure to use ``clang`` (not ``ld``) for the final link step. To
get a reasonable performance add ``-O1`` or higher.
TypeSanitizer by default doesn't print the full stack trace on error messages. Use ``TYSAN_OPTIONS=print_stacktrace=1`` 
to print the full trace. To get nicer stack traces in error messages add ``-fno-omit-frame-pointer`` and 
``-g``.  To get perfect stack traces you may need to disable inlining (just use ``-O1``) and tail call elimination 
(``-fno-optimize-sibling-calls``).

.. code-block:: console

    % cat example_AliasViolation.c
    int main(int argc, char **argv) {
      int x = 100;
      float *y = (float*)&x;
      *y += 2.0f;          // Strict aliasing violation
      return 0;
    }

    # Compile and link
    % clang++ -g -fsanitize=type example_AliasViolation.cc

> If a strict aliasing violation is detected, the program will print an error message to stderr. 

Reviewer: ```suggestion
The program will print an error message to stderr each time a strict aliasing violation is detected. 
```

The program won't terminate, which will allow you to detect many strict aliasing violations in one 
run.

.. code-block:: console

    % ./a.out
    ==1375532==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7ffeebf1a72c (pc 0x5b3b1145ff41 bp 0x7ffeebf1a660 sp 0x7ffeebf19e08 tid 1375532)
    READ of size 4 at 0x7ffeebf1a72c with type float accesses an existing object of type int
        #0 0x5b3b1145ff40 in main example_AliasViolation.c:4:10

    ==1375532==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7ffeebf1a72c (pc 0x5b3b1146008a bp 0x7ffeebf1a660 sp 0x7ffeebf19e08 tid 1375532)
    WRITE of size 4 at 0x7ffeebf1a72c with type float accesses an existing object of type int
        #0 0x5b3b11460089 in main example_AliasViolation.c:4:10

Error terminology
------------------

There are some terms that may appear in TypeSanitizer errors that are derived from 
`TBAA Metadata <https://llvm.org/docs/LangRef.html#tbaa-metadata>`. This section hopes to provide a 
brief dictionary of these terms.

* ``omnipotent char``: This is a special type which can alias with anything. Its name comes from the C/C++ 
  type ``char``.

> * ``type p[x]``: Sometimes a program could generate distinct TBAA metadata that resolve to the same name. 

Reviewer: `p2 int` means `int **`, the number indicates the number of indirections (not sure what the best terminology would be here

Author: Ah! My mistake, will change that

  To make them unique, they have the character 'p' and a number prepended to their name.

These terms are a result of non-user-facing processes, and not always self-explanatory. There is some 

> interest in changing TypeSanitizer in the future to translate these terms before printing them to users.

Reviewer: Might be worth just saying that it is still experimental and the user-facing error messages should be improved in the future to remove LLVM IR specific references

function attribute ``no_sanitize("type")`` to disable instrumenting type aliasing. 
It is possible, depending on what happens in non-instrumented code, that instrumented code 
emits false-positives/ false-negatives. This attribute may not be supported by other 
compilers, so we suggest to use it together with ``__has_feature(type_sanitizer)``.

``__attribute__((disable_sanitizer_instrumentation))``
--------------------------------------------------------

The ``disable_sanitizer_instrumentation`` attribute can be applied to functions
to prevent all kinds of instrumentation. As a result, it may introduce false
positives and incorrect stack traces. Therefore, it should be used with care,
and only if absolutely required; for example for certain code that cannot
tolerate any instrumentation and resulting side-effects. This attribute
overrides ``no_sanitize("type")``.

Ignorelist
----------

TypeSanitizer supports ``src`` and ``fun`` entity types in
:doc:`SanitizerSpecialCaseList`, that can be used to suppress aliasing 
violation reports in the specified source files or functions. Like 
with other methods of ignoring instrumentation, this can result in false 

> positives/ false-negatives.

Reviewer: I am not sure if anyone ever tried this? Do we have test with an ignore list?

Author: We don't have these as tysan tests, but I did test this and it worked. Same with the other sanitizer features.
Is it worth me putting these tests up in another PR?

Reviewer 3: Yes please, we should have test coverage for anything user-facing.

Author: Opened PR #124125

Limitations
-----------

* TypeSanitizer uses more real memory than a native run. It uses 8 bytes of
  shadow memory for each byte of user memory.
* There are transformation passes which run before TypeSanitizer. If these 
  passes optimize out an aliasing violation, TypeSanitizer cannot catch it.
* Currently, all instrumentation is inlined. This can result in a **15x** 
  (on average) increase in generated file size, and **3x** to **7x** increase 
  in compile time. In some documented cases this can cause the compiler to hang.
  There are plans to improve this in the future.
* Codebases that use unions and struct-initialized variables can see incorrect 
  results, as TypeSanitizer doesn't yet instrument these reliably.

Current Status
--------------

TypeSanitizer is brand new, and still in development. There are some known 

> issues, especially in areas where clang doesn't generate valid TBAA metadata. 

Reviewer: ```suggestion
issues, especially in areas where Clang doesn't generate valid TBAA metadata. 
```

I am not sure saying Clang doesn't generate valid TBAA metadata is correct here. It may not generate TBAA metadata that TySan suspects in some cases, but what Clang emits should be conservatively correct.

Author: Aha right, its not that it isn't valid, it just isn't descriptive/extensive enough for the sanitizer. Will change that

> We are actively working on enhancing the tool --- stay tuned.  Any help, 

Reviewer: Might be good to link to the issue tracker and possibly mention the tag for tysan issues.

================
TypeSanitizer
================

.. contents::
   :local:

Introduction
============

> TypeSanitizer is a detector for strict type aliasing violations. It consists of a compiler

Reviewer 4: ```suggestion
The TypeSanitizer is a detector for strict type aliasing violations. It consists of a compiler
```

> instrumentation module and a run-time library. The tool detects violations where you access 

Reviewer 4: `violations` of what?  Would love another sentence or two here just explaining what it is detecting better.  Something like (please don't use unless it is correct!):

"This tool detects violations of the strict-aliasing rule, which prohibits access of a memory location as a type that is different from the dynamic type of the object at that location."

Or something?

Author: I added some more text based off Florian's [original pitch](https://reviews.llvm.org/D32198)

memory under a different type than the dynamic type of the object.

The violations TypeSanitizer catches may cause the compiler to emit incorrect code.

As TypeSanitizer is still experimental, it can currently have a large impact on runtime speed, 
memory use, and code size.

How to build
============

Build LLVM/Clang with `CMake <https://llvm.org/docs/CMake.html>`_ and enable
the ``compiler-rt`` runtime. An example CMake configuration that will allow
for the use/testing of TypeSanitizer:

.. code-block:: console

   $ cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_RUNTIMES="compiler-rt" <path to source>/llvm

Usage
=====

Compile and link your program with ``-fsanitize=type`` flag.  The
TypeSanitizer run-time library should be linked to the final executable, so
make sure to use ``clang`` (not ``ld``) for the final link step. To
get a reasonable performance add ``-O1`` or higher.

> TypeSanitizer by default doesn't print the full stack trace on error messages. Use ``TYSAN_OPTIONS=print_stacktrace=1`` 

Reviewer 4: ```suggestion
TypeSanitizer by default doesn't print the full stack trace in error messages. Use ``TYSAN_OPTIONS=print_stacktrace=1`` 
```

> ================
> TypeSanitizer
> ================

Reviewer 3: ```suggestion
=============
TypeSanitizer
=============
```

can exploit these for optimizations given the TBAA metadata Clang emits. In general, a pointer 
of a given type cannot access an object of a different type, with only a few exceptions. 

These rules aren't always apparent to users, which leads to code that violates these rules
(e.g. for type punning). This can lead to optimization passes introducing bugs unless the 
code is build with ``-fno-strict-aliasing``, sacrificing performance.

TypeSanitizer is built to catch when these strict aliasing rules have been violated, helping 
users find where such bugs originate in their code despite the code looking valid at first glance.

As TypeSanitizer is still experimental, it can currently have a large impact on runtime speed, 
memory use, and code size.

How to build
============

Build LLVM/Clang with `CMake <https://llvm.org/docs/CMake.html>`_ and enable
the ``compiler-rt`` runtime. An example CMake configuration that will allow
for the use/testing of TypeSanitizer:

.. code-block:: console

   $ cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_PROJECTS="clang" -DLLVM_ENABLE_RUNTIMES="compiler-rt" <path to source>/llvm

Usage
=====

> Compile and link your program with ``-fsanitize=type`` flag.  The

Reviewer 3: Can the type sanitizer be enabled along with other sanitizers, or is it mutually exclusive with some? (If it's mutually exclusive with some, then we should document which ones, otherwise I think the docs are fine as-is.)

Author: From a check, I think its mutually exclusive with all of them currently. I will put this in the **Limitations** section

TypeSanitizer run-time library should be linked to the final executable, so
make sure to use ``clang`` (not ``ld``) for the final link step. To
get a reasonable performance add ``-O1`` or higher.
TypeSanitizer by default doesn't print the full stack trace in error messages. Use ``TYSAN_OPTIONS=print_stacktrace=1`` 
to print the full trace. To get nicer stack traces in error messages add ``-fno-omit-frame-pointer`` and 
``-g``.  To get perfect stack traces you may need to disable inlining (just use ``-O1``) and tail call elimination 
(``-fno-optimize-sibling-calls``).

.. code-block:: console

    % cat example_AliasViolation.c
    int main(int argc, char **argv) {
      int x = 100;
      float *y = (float*)&x;
      *y += 2.0f;          // Strict aliasing violation
      return 0;
    }

    # Compile and link
    % clang++ -g -fsanitize=type example_AliasViolation.cc

> The program will print an error message to stderr each time a strict aliasing violation is detected. 

Reviewer 3: ```suggestion
The program will print an error message to ``stderr`` each time a strict aliasing violation is detected. 
```

The program won't terminate, which will allow you to detect many strict aliasing violations in one 
run.

.. code-block:: console

    % ./a.out
    ==1375532==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7ffeebf1a72c (pc 0x5b3b1145ff41 bp 0x7ffeebf1a660 sp 0x7ffeebf19e08 tid 1375532)
    READ of size 4 at 0x7ffeebf1a72c with type float accesses an existing object of type int
        #0 0x5b3b1145ff40 in main example_AliasViolation.c:4:10

    ==1375532==ERROR: TypeSanitizer: type-aliasing-violation on address 0x7ffeebf1a72c (pc 0x5b3b1146008a bp 0x7ffeebf1a660 sp 0x7ffeebf19e08 tid 1375532)
    WRITE of size 4 at 0x7ffeebf1a72c with type float accesses an existing object of type int
        #0 0x5b3b11460089 in main example_AliasViolation.c:4:10

Error terminology
------------------

There are some terms that may appear in TypeSanitizer errors that are derived from 
`TBAA Metadata <https://llvm.org/docs/LangRef.html#tbaa-metadata>`. This section hopes to provide a 
brief dictionary of these terms.

* ``omnipotent char``: This is a special type which can alias with anything. Its name comes from the C/C++ 
  type ``char``.

> * ``type p[x]``: This signifies pointers to the type. x is the number of indirections to reach the final value.

Reviewer 3: ```suggestion
* ``type p[x]``: This signifies pointers to the type. ``x`` is the number of indirections to reach the final value.
```

It is possible, depending on what happens in non-instrumented code, that instrumented code 
emits false-positives/ false-negatives. This attribute may not be supported by other 
compilers, so we suggest to use it together with ``__has_feature(type_sanitizer)``.

``__attribute__((disable_sanitizer_instrumentation))``
--------------------------------------------------------

The ``disable_sanitizer_instrumentation`` attribute can be applied to functions
to prevent all kinds of instrumentation. As a result, it may introduce false
positives and incorrect stack traces. Therefore, it should be used with care,
and only if absolutely required; for example for certain code that cannot
tolerate any instrumentation and resulting side-effects. This attribute
overrides ``no_sanitize("type")``.

Ignorelist
----------

TypeSanitizer supports ``src`` and ``fun`` entity types in
:doc:`SanitizerSpecialCaseList`, that can be used to suppress aliasing 
violation reports in the specified source files or functions. Like 
with other methods of ignoring instrumentation, this can result in false 
positives/ false-negatives.

> Limitations

Reviewer 3: One thing we may want to document is whether type sanitizer follows Clang's & LLVM's rules for type aliasing or the standard's rules for type aliasing. e.g., this is valid type punning in C, but not in C++:
```
static_assert(sizeof(int) == sizeof(float));
union U {
  int i;
  float f;
};

union U u;
u.i = 12;
float f = u.f;
```
IIRC, Clang and LLVM don't always agree with the standard on what is or isn't allowed in these cases.

The TypeSanitizer is a detector for strict type aliasing violations. It consists of a compiler
instrumentation module and a run-time library. C/C++ has type-based aliasing rules, and LLVM 
can exploit these for optimizations given the TBAA metadata Clang emits. In general, a pointer 
of a given type cannot access an object of a different type, with only a few exceptions. 

These rules aren't always apparent to users, which leads to code that violates these rules
(e.g. for type punning). This can lead to optimization passes introducing bugs unless the 
code is build with ``-fno-strict-aliasing``, sacrificing performance.

TypeSanitizer is built to catch when these strict aliasing rules have been violated, helping 
users find where such bugs originate in their code despite the code looking valid at first glance.

As TypeSanitizer is still experimental, it can currently have a large impact on runtime speed, 
memory use, and code size. It also has a large compile-time overhead. Work is being done to 
reduce these impacts.

The TypeSanitizer Algorithm
===========================
For each TBAA type-access descriptor, encoded in LLVM IR using TBAA Metadata, the instrumentation 

> pass generates descriptor tales. Thus there is a unique pointer to each type (and access descriptor).

Reviewer 5: Small typo: s/tales/tables/

Author: Thanks, will fix!
